# ME 581 HW 2 Solution [Python3]
Yi Yang

## Problem 1
A system of equation $Ax = b$ is given as
$$\begin{bmatrix}
2.01 & 1.99\\
1.99 & 2.01
\end{bmatrix}
x = 
\begin{bmatrix}
4\\
4
\end{bmatrix}.$$
The Solution to this system of equations is 
$$x = 
\begin{bmatrix}
1\\
1
\end{bmatrix},$$
and one approximate solution is 
$$\tilde{x} = 
\begin{bmatrix}
2\\
0
\end{bmatrix}.$$
(i) Compute the error $e = \tilde{x} - x$.

(ii) Compute the residual $r = A\tilde{x} - b$.

(iii) Using the $l_{\infty}$ norm, compute the relative error $\tfrac{\|e\|_{\infty}}{\|x\|_{\infty}}$.

(iv) Using the $l_{\infty}$ norm, compute the condition number $\kappa_{\infty}$. 

(v) Using the $l_{\infty}$ norm, compute the relative residual $\tfrac{\|r\|_{\infty}}{\|b\|_{\infty}}$.

(vi) Compute the product of the condition number and the relative residual.

(vii) Compare the relative error to the product of the condition number and the relative residual.

**Solution: ** 

(i) The error is 
$$e = \tilde{x} - x = \begin{bmatrix}
1\\
-1
\end{bmatrix}$$.

(ii) The residual is:
$$r = A\tilde{x} - b = \begin{bmatrix}
2.01 & 1.99\\
1.99 & 2.01
\end{bmatrix}\begin{bmatrix}
2\\
0
\end{bmatrix} - \begin{bmatrix}
4\\
4
\end{bmatrix} = 
\begin{bmatrix}
0.02\\
-0.02
\end{bmatrix}$$

(iii) The relative error is:
$$\frac{\|e\|_{\infty}}{\|x\|_{\infty}} = \frac{\max_{i} |e_i|}{\max_{i} |x_i|} = \frac{1}{1} = 1$$

(iv) The coefficient matrix A and inverse of matrix A are given below:
$$A = \begin{bmatrix}
2.01 & 1.99\\
1.99 & 2.01
\end{bmatrix},\quad
A^{-1} = \begin{bmatrix}
25.1250 & -24.8750\\
-24.8750 & 25.1250
\end{bmatrix}$$
Therefore, the condition number is:
$$\kappa_{\infty} = \|A\|_{\infty}\cdot\|A^{-1}\|_{\infty} = 4\times 50 = 200$$

(v) The relative residual is:
$$\frac{\|r\|_{\infty}}{\|b\|_{\infty}} = \frac{0.02}{4} = 0.005$$

(vi) The product of the condition number and the relative residual:
$$\kappa_{\infty}\times \frac{\|r\|_{\infty}}{\|b\|_{\infty}} = 200 \times 0.005 = 1$$

(vii) The relative error are equal to the product of the condition number and the relative residual.

## Problem 2
Let
$$A = \begin{bmatrix}
5.1 & 8.7\\
2.4 & 4.1
\end{bmatrix} \text{and } b = \begin{bmatrix}
9.48\\
4.48
\end{bmatrix}.$$
(i) Using the $l_{\infty}$ norm, compute the condition number $\kappa_{\infty}(A)$.

(ii) Solve the system of equations $Ax = b$ for $x$.

(iii) Perturb the coefficient matrix $A$ and the right-side vector $b$ by
$$\delta A = \begin{bmatrix}
-0.001 & 0\\
0.001 & 0
\end{bmatrix} \text{and } \delta b = \begin{bmatrix}
0.05\\
-0.05
\end{bmatrix}$$
and solve the resulting perturbed system of equations $(A + \delta A)\tilde{x} = (b + \delta b)$ for the approximate solution $\tilde{x}$.

(iv) Using the $l_{\infty}$ norm, compute the actual value of the relative change in the solution, $\tfrac{\|\delta x\|_{\infty}}{\|x\|_{\infty}}$ for the perturbation in the part (iii).

(v) Using the $l_\infty$ norm, compute the theoretical upper bound of the relative change in the solution, $\tfrac{\|\delta x\|_{\infty}}{\|x\|_{\infty}}$ for the perturbation in part (iii).

(vi) For the perturbation in part (iii), compare the actual value of $\tfrac{\|\delta x\|_{\infty}}{\|x\|_{\infty}}$ to its theoretical upper bound.

(vii) Perturb the original coefficient matrix $A$ and the original right-side vector $b$ by
$$\delta A = \begin{bmatrix}
0.001 & -0.001\\
-0.001 & 0.001
\end{bmatrix} \text{and } \delta b = \begin{bmatrix}
-0.1\\
0.1
\end{bmatrix}$$
and solve the resulting perturbed system of equations $(A + \delta A)\tilde{x} = (b + \delta b)$ for $\tilde{x}$.

(viii) Using the $l_\infty$ norm, compute the actual value of the relative change in the solution, $\tfrac{\|\delta x\|_{\infty}}{\|x\|_{\infty}}$ for the perturbation in part (vii).

(ix) Using the $l_\infty$ norm, compute the theoretical upper bound of the relative change in the solution, $\tfrac{\|\delta x\|_{\infty}}{\|x\|_{\infty}}$ for the perturbation in part (vii).

(x) For the perturbation in part (vii), compare the actual value of $\tfrac{\|\delta x\|_{\infty}}{\|x\|_{\infty}}$ to its theoretical upper bound.

**Solution:** 

(i) Matrix A and $A^{-1}$ are shown below:
$$A = \begin{bmatrix}
5.1 & 8.7\\
2.4 & 4.1
\end{bmatrix} \text{and } A^{-1} = \begin{bmatrix}
\frac{6970}{51} & -290\\
-80 & 170
\end{bmatrix}$$
Therefore, the condition number is 
$$\kappa_\infty = \|A\|_{\infty}\cdot\|A^{-1}\|_{\infty} = 13.8\times (426 + \frac{34}{51}) = 5888$$

(ii) Using Gauss Elimination to solve for this system of equations, first forward elimination I can get:
$$\begin{bmatrix}
17 & 29\\
0 & \frac{1}{51}
\end{bmatrix}x = \begin{bmatrix}
31.6\\
\frac{3.2}{51}
\end{bmatrix}$$
Next, back subsitution to get:
$$x_2 = 3.2,\quad x_1 = (31.6 - 29\times 3.2)/17 = -3.6$$
Therefore, solve for the system of equations I can get
$$x = \begin{bmatrix}
-3.6\\
3.2
\end{bmatrix}$$

(iii) The perturbed system of equations $(A + \delta A)\tilde{x} = (b + \delta b)$ is shown below:
$$\begin{bmatrix}
5.099 & 8.700\\
2.401 & 4.100
\end{bmatrix}\tilde{x} = \begin{bmatrix}
9.53\\
4.43
\end{bmatrix}$$
The approximate solution is 
$$\tilde{x} = \begin{bmatrix}
30.93023\\
-17.03256
\end{bmatrix}$$

(iv) $\delta x$ and $x$ are given below:
$$x = \begin{bmatrix}
-3.6\\
3.2
\end{bmatrix}\quad \delta x = \begin{bmatrix}
34.53023\\
-20.23256
\end{bmatrix}$$
The actual value of relative change is 
$$\frac{\|\delta x\|_{\infty}}{\|x\|_{\infty}} = \frac{34.53023}{3.6} = 9.5917$$

(v) Theorem: Assume $Ax = b$ and $\hat{A}\hat{x} = \hat{b}$. If $A$ is nonsingular and $\|\hat{A} - A\| < \frac{1}{\|A^{-1}\|}$, there holds
$$\frac{\|\hat{x} - x\|}{\|x\|} \leq \frac{cond(A)}{1-cond(A)\frac{\|\hat{A} - A\|}{\|A\|}}(\frac{\|\hat{b}-b\|}{\|b\|} + \frac{\|\hat{A} - A\|}{\|A\|})$$
Proof: Suppose $E = \hat{A} - A$, and we have $(A+E)\hat{x} = \hat{b}$, $Ax = b$, then
$$A(\hat{x} - x) = \hat{b} - b - E\hat{x}$$
$$\|\hat{x} - x\| \leq \|A^{-1}\|(\|\hat{b} - b\| + \|E\|\cdot\|\hat{x}\|) 
\leq \|A^{-1}\|\left[\|\hat{b} - b\| + \|E\|\cdot (\|\hat{x} - x\| + \|x\|)\right]$$
Therefore,
$$(1 - \|A^{-1}\|\cdot\|E\|)\|\hat{x} - x\| \leq \|A^{-1}\|(\|\hat{b}-b\| + \|E\|\|x\|)$$
Since $1 \leq \tfrac{\|A\|\|x\|}{\|b\|}$ and $\|\hat{A} - A\| < \frac{1}{\|A^{-1}\|}$,
$$(1 - \|A^{-1}\|\cdot\|E\|)\|\hat{x} - x\| \leq \|A^{-1}\|(\frac{\|\hat{b}-b\|}{\|b\|}\cdot \|A\|\|x\| + \|E\|\|x\|)$$
Hence, the theorem is proved:
$$\frac{\|\hat{x} - x\|}{\|x\|} \leq \frac{cond(A)}{1-cond(A)\frac{\|\hat{A} - A\|}{\|A\|}}(\frac{\|\hat{b}-b\|}{\|b\|} + \frac{\|\hat{A} - A\|}{\|A\|})$$
The theoretical upper bound of the relative change in the solution is:
$$\frac{\|\delta x\|_{\infty}}{\|x\|_{\infty}} \leq \frac{cond(A)}{1-cond(A)\frac{\|\hat{A} - A\|}{\|A\|}}(\frac{\|\hat{b}-b\|}{\|b\|} + \frac{\|\hat{A} - A\|}{\|A\|}) = \frac{5888}{1 - 5888\times \frac{0.001}{13.8}}\left(\frac{0.05}{9.48} + \frac{0.001}{13.8}\right) = 54.9096$$

(vi) For the perturbation in part (iii), the actual value of $\tfrac{\|\delta x\|_{\infty}}{\|x\|_{\infty}} = 9.5917$ is less than upper bound which is $54.9096$. 

(vii) The new perturbed system of equations are shown below:
$$\begin{bmatrix}
5.101 & 8.699\\
2.399 & 4.101
\end{bmatrix}\tilde{x} = \begin{bmatrix}
9.38\\
4.58
\end{bmatrix}$$
The approximate solution is 
$$\tilde{x} = \begin{bmatrix}
-27.3169\\
17.0966
\end{bmatrix}$$

(viii) $\delta x$ and $x$ are given below:
$$x = \begin{bmatrix}
-3.6\\
3.2
\end{bmatrix}\quad \delta x = \begin{bmatrix}
-23.7169\\
13.8766
\end{bmatrix}$$
The actual value of relative change is 
$$\frac{\|\delta x\|_{\infty}}{\|x\|_{\infty}} = \frac{23.7169}{3.6} = 6.588$$

(ix) The theoretical upper bound of the relative change in the solution is:
$$\frac{\|\delta x\|_{\infty}}{\|x\|_{\infty}} \leq \frac{cond(A)}{1-cond(A)\frac{\|\hat{A} - A\|}{\|A\|}}(\frac{\|\hat{b}-b\|}{\|b\|} + \frac{\|\hat{A} - A\|}{\|A\|}) = \frac{5888}{1 - 5888\times \frac{0.002}{13.8}}\left(\frac{0.1}{9.48} + \frac{0.002}{13.8}\right) = 429.2934$$

(x) The actual value of $\tfrac{\|\delta x\|_{\infty}}{\|x\|_{\infty}} = 6.588$ is smaller than the theoretical upper bound, which is $429.2934$.

## Problem 3
Solve the augmented matrix
$$\begin{bmatrix}
3 & 1 & 4 & -1 & 7\\
2 & -2 & -1 & 2 & 1\\
5 & 7 & 14 & -9 & 21\\
1 & 3 & 2 & 4 & -4
\end{bmatrix}$$
By means of 

(i) Gaussian Elimination with Partial Pivoting.

(ii) Gaussian Elimination with scaled Partial Pivoting.

**Solution:**

(i) The Gaussian Elimination with Partial Pivoting is shown below:
$$\begin{bmatrix}
3 & 1 & 4 & -1 & 7\\
2 & -2 & -1 & 2 & 1\\
5 & 7 & 14 & -9 & 21\\
1 & 3 & 2 & 4 & -4
\end{bmatrix} \rightarrow \begin{bmatrix}
5 & 7 & 14 & -9 & 21\\
0 & -\frac{24}{5} & -\frac{33}{5} & \frac{28}{5} & -\frac{37}{5}\\
0 & -\frac{16}{5} & -\frac{22}{5} & \frac{22}{5} & -\frac{28}{5}\\
0 & \frac{8}{5} & -\frac{4}{5} & \frac{29}{5} & -\frac{41}{5}
\end{bmatrix} \rightarrow \begin{bmatrix}
5 & 7 & 14 & -9 & 21\\
0 & -\frac{24}{5} & -\frac{33}{5} & \frac{28}{5} & -\frac{37}{5}\\
0 & 0 & 0 & \frac{2}{3} & -\frac{2}{3}\\
0 & 0 & -3 & \frac{23}{3} & -\frac{32}{3}
\end{bmatrix}$$
$$ \rightarrow \begin{bmatrix}
5 & 7 & 14 & -9 & 21\\
0 & -\frac{24}{5} & -\frac{33}{5} & \frac{28}{5} & -\frac{37}{5}\\
0 & 0 & -3 & \frac{23}{3} & -\frac{32}{3}\\
0 & 0 & 0 & \frac{2}{3} & -\frac{2}{3}
\end{bmatrix}$$
Using back substitution, we can get the final result
$$x_4 = -\frac{2}{3}/\frac{2}{3} = -1, \quad x_3 = (-\frac{32}{3} - \frac{23}{3}\times x_4)/(-3) = 1, \dots$$
The final solution using Gaussian Elimination with Partial Pivoting is:
$$x = \begin{bmatrix}
1\\
-1\\
1\\
-1
\end{bmatrix}$$

(ii) Scaled partial pivoting, we choose as pivot equation an equation from the $n - k + 1$ available equations such that the quotient of the coefficient of $x_k$ and the absolutely largest coefficient in the equation is maximum in absolute value. If there are several such equations, take the first of them. We have 
$$\frac{|a_{11}|}{\max|a_{1k}|} = \frac{3}{4} \quad \frac{|a_{21}|}{\max|a_{2k}|} = 1 \quad \frac{|a_{31}|}{\max|a_{3k}|} = \frac{5}{14} \quad \frac{|a_{41}|}{\max|a_{4k}|} = \frac{1}{4}$$
Hence, we interchange row 1 and row 2 and apply forward elimination. And we get:
$$\begin{bmatrix}
2 & -2 & -1 & 2 & 1\\
0 & 4 & \frac{11}{2} & -4 & \frac{11}{2}\\
0 & 12 & \frac{33}{2} & -14 & \frac{37}{2}\\
0 & 4 & \frac{5}{2} & 3 & -\frac{9}{2}
\end{bmatrix}$$
Then, for row 2, row 3 and row 4 we have:
$$\frac{|a_{22}|}{\max|a_{2k}|} = \frac{8}{11} \quad \frac{|a_{32}|}{\max|a_{3k}|} = \frac{24}{33} \quad \frac{|a_{42}|}{\max|a_{4k}|} = 1$$
Hence, we interchange row 2 and row 4 and apply forward elimination. And we get:
$$\begin{bmatrix}
2 & -2 & -1 & 2 & 1\\
0 & 4 & \frac{5}{2} & 3 & -\frac{9}{2}\\
0 & 0 & 3 & -7 & 10\\
0 & 0 & 9 & -23 & 32
\end{bmatrix}$$
Last, we operate on row 3 and row 4 we have:
$$\frac{|a_{33}|}{\max|a_{3k}|} = \frac{3}{7} \quad \frac{|a_{43}|}{\max|a_{4k}|} = \frac{9}{23}$$
Hence, we remain the order of last two rows and apply forward elimination. And we get:
$$\begin{bmatrix}
2 & -2 & -1 & 2 & 1\\
0 & 4 & \frac{5}{2} & 3 & -\frac{9}{2}\\
0 & 0 & 3 & -7 & 10\\
0 & 0 & 0 & -2 & 2
\end{bmatrix}$$
Apply back substitution to the last form of augmented matrix, we have:
$$x = \begin{bmatrix}
1\\
-1\\
1\\
-1
\end{bmatrix}$$

## Problem 4
(i) Solve the augmented matrix by means of Gaussian Elimination with Partial Pivoting in double precision:
$$\begin{bmatrix}
-9 & 11 & -21 & 63 & -252 & -356\\
70 & -69 & 141 & -421 & 1684 & 2385\\
-575 & 575 & -1149 & 3451 & -13801 & -19551\\
3891 & -3891 & 7782 & -23345 & 93365 & 132274\\
1024 & -1024 & 2048 & -6144 & 24572 & 34812
\end{bmatrix}$$

(ii) Using the $l_\infty$ norm, estimate the condition number of the coefficient matrix based on your result. The exact solution for this problem is given by $x = \begin{bmatrix}
1 & -1 & 1 & -1 & 1
\end{bmatrix}^T$.

**Solution:**
 
(i) The codes for solving this system of linear equations using Gaussian Elimination with Partial Pivoting is shown below:

In [28]:
import numpy as np
from matplotlib import pyplot as plt
# Implementation of Gaussian Elimination with Partial Pivoting
# This algorithm computes a unique solution of the system, or 
# indicate that system has no unique solution
# Partial Pivoting when Akk is less than 0.1% of maximum in cl
def gaussElim(A, b):
    n = len(b)
    for k in range(0,n-1):
        '''
        maximum = abs(A[k,k])
        r = k
        for i in range(k+1,n):
            if abs(A[i,k]) > maximum:
                maximum = abs(A[i,k])
                r = i
        '''
        r = abs(A[k:,k]).argmax() + k
        if A[r,k] == 0.0:
            print("The system does not have unique solution.")
            break
        else:
            '''
            tmp = A[r,:].copy()
            tmpb = b[r].copy()
            A[r,:] = A[k,:]
            b[r] = b[k]
            A[k,:] = tmp
            b[k] = tmpb
            '''
            A[[k,r]] = A[[r,k]]
            b[[k,r]] = b[[r,k]]
        for i in range(k+1,n):
            lam = A[i,k]/A[k,k]
            A[i,k+1:n] = A[i,k+1:n] - lam*A[k,k+1:n]
            b[i] = b[i] - lam*b[k]
    b[n-1] = b[n-1]/A[n-1,n-1]
    for k in range(n-2,-1,-1):
        b[k] = (b[k] - A[k,k+1:n].dot(b[k+1:n]))/A[k,k]
    return b

A = np.array([[-9.0, 11.0, -21.0, 63.0, -252.0],\
              [70.0, -69.0, 141.0, -421.0, 1684.0], \
              [-575.0, 575.0, -1149.0, 3451.0, -13801.0],\
              [3891.0, -3891.0, 7782.0, -23345.0, 93365.0],\
              [1024.0, -1024.0, 2048.0, -6144.0, 24572.0]])
b = np.array([-356.0, 2385.0, -19551.0, 132274.0, 34812.0])
# !!note function paremeters are passed by reference here!!
x = gaussElim(A.copy(),b.copy())
print("***ME 581 Homework 2 Print Out***\n")
print("The solution from Gaussian Elimination with Partial Pivoting is ", x)

***ME 581 Homework 2 Print Out***

The solution from Gaussian Elimination with Partial Pivoting is  [ 1.         -0.86468439  0.0850087   5.26962347  2.64956178]


(ii) Using Gaussian Elimination with Partial Pivoting, we get the solution is $$\hat{x} = \begin{bmatrix}
1. & -0.86468439 & 0.0850087 & 5.26962347 & 2.64956178
\end{bmatrix}^T.$$
The estimate condition number is shown below:

In [29]:
x0 = np.array([1.0, -1.0, 1.0, -1.0, 1.0])
deltax = x - x0
deltab = b - A.dot(x)
kappa = (np.linalg.norm(deltax,np.inf)/1.0)/\
(np.linalg.norm(deltab,np.inf)/np.linalg.norm(b,np.inf))
print("***ME 581 Homework 2 Print Out***\n")
print("Estimate Condition number: ", kappa)
print("Exact Condition number: ", np.linalg.cond(A,np.inf))
print(A.dot(x))
print(np.linalg.inv(A).dot(b))

***ME 581 Homework 2 Print Out***

Estimate Condition number:  2.27958495215e+17
Exact Condition number:  4.4062629868e+17
[   -356.    2385.  -19551.  132274.   34812.]
[-19.07776585 -12.           4.         -16.           0.        ]


## Problem 5 
Determine the member and reaction forces within the plane truss shown in Figure 1 when the truss is subjected to each of the following loading configurations.

(a) 500-pound forces directed vertically downward at nodes #3 and #5, and a 1000-pound force directed vertically downward at node #4.

(b) A 500-pound force directed vertically downward at nodes #3, a 1000-pound force directed vertically downward at node #4, a 1500-pound force directed vertically downward at node #5.

(c) A 1500-pound force directed downward at nodes #3, a 1000-pound force directed vertically downward at node #4, a 500-pound force directed vertically downward at node #5.

(d) 500-pound force acting at node #4, and a 1000-pound force acting at node #3, both forces acting horizontally to the right.

(e) 500-pound force acting at node #4, and a 1000-pound force acting at node #5, both forces acting horizontally to the left.

Solve the problem using your GE code with partial pivoting. Show the augmented matrix and the resulting forces for each case.

![Figure1](me582hw2pb5.png)

**Solution:**

(a) For Node #1,
$$\frac{1}{\sqrt{5}}F_1 + \frac{4}{5}F_3 + F_V = 0$$
$$\frac{2}{\sqrt{5}}F_1 + \frac{3}{5}F_3 + F_H = 0$$
For Node #6,
$$\frac{3}{5}F_7 + \frac{1}{\sqrt{5}}F_9 = 0$$
$$\frac{4}{5}F_7 + \frac{2}{\sqrt{5}}F_9 + F_R = 0$$
For Node #3,
$$\frac{3}{5}F_3 - \frac{3}{\sqrt{10}}F_2 - \frac{3}{\sqrt{10}}F_4 = 0$$
$$\frac{4}{5}F_3 + \frac{1}{\sqrt{10}}F_2 - \frac{1}{\sqrt{10}}F_4 + P_3 = 0$$
For Node #4,
$$\frac{3}{\sqrt{10}}F_4 - \frac{3}{\sqrt{10}}F_6 = 0$$
$$\frac{1}{\sqrt{10}}F_4 + \frac{1}{\sqrt{10}}F_6 + F_5 + P_4 = 0$$
For Node #5,
$$\frac{3}{\sqrt{10}}F_6 + \frac{3}{\sqrt{10}}F_8 - \frac{3}{5}F_7 = 0$$
$$\frac{1}{\sqrt{10}}F_8 + \frac{4}{5}F_7 - \frac{1}{\sqrt{10}}F_6 + P_5 = 0$$
For Node #2,
$$\frac{3}{\sqrt{10}}F_2 + \frac{2}{\sqrt{5}}F_1 - \frac{3}{\sqrt{10}}F_8 - \frac{2}{\sqrt{5}}F_9 = 0$$
$$\frac{1}{\sqrt{5}}F_1 + \frac{1}{\sqrt{5}}F_9 - \frac{1}{\sqrt{10}}F_2 - \frac{1}{\sqrt{10}}F_8 - F_5 = 0$$
The augmented matrix is:
$$\tilde{A} = \begin{bmatrix}
\frac{1}{\sqrt{5}} & 0 & \frac{4}{5} & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0\\
\frac{2}{\sqrt{5}} & 0 & \frac{3}{5} & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0\\
0 & 0 & 0 & 0 & 0 & 0 & \frac{3}{5} & 0 & \frac{1}{\sqrt{5}} & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 0 & 0 & 0 & \frac{4}{5} & 0 & \frac{2}{\sqrt{5}} & 0 & 0 & 1 & 0\\
0 & -\frac{3}{\sqrt{10}} & \frac{3}{5} & -\frac{3}{\sqrt{10}} & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\
0 & \frac{1}{\sqrt{10}} & \frac{4}{5} & -\frac{1}{\sqrt{10}} & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -P_3\\
0 & 0 & 0 & \frac{3}{\sqrt{10}} & 0 & -\frac{3}{\sqrt{10}} & 0 & 0 & 0 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & \frac{1}{\sqrt{10}} & 1 & \frac{1}{\sqrt{10}} & 0 & 0 & 0 & 0 & 0 & 0 & -P_4\\
0 & 0 & 0 & 0 & 0 & \frac{3}{\sqrt{10}} & -\frac{3}{5} & \frac{3}{\sqrt{10}} & 0 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 0 & 0 & -\frac{1}{\sqrt{10}} & \frac{4}{5} & \frac{1}{\sqrt{10}} & 0 & 0 & 0 & 0 & -P_5\\
\frac{2}{\sqrt{5}} & \frac{3}{\sqrt{10}} & 0 & 0 & 0 & 0 & 0 & -\frac{3}{\sqrt{10}} & -\frac{2}{\sqrt{5}} & 0 & 0 & 0 & 0\\
\frac{1}{\sqrt{5}} & -\frac{1}{\sqrt{10}} & 0 & 0 & -1 & 0 & 0 & -\frac{1}{\sqrt{10}} & \frac{1}{\sqrt{5}} & 0 & 0 & 0 & 0
\end{bmatrix}$$
The code for solving this system using Gaussian Elimination with Partial Pivoting is shown below:

In [30]:
A51 = np.array([[1./np.sqrt(5.), 0., 4./5., 0., 0., 0., 0., 0., 0., 1., 0., 0.],\
                [2./np.sqrt(5.), 0., 3./5., 0., 0., 0., 0., 0., 0., 0., 1., 0.], \
                [0., 0., 0., 0., 0., 0., 3./5., 0., 1./np.sqrt(5.), 0., 0., 0.], \
                [0., 0., 0., 0., 0., 0., 4./5., 0., 2./np.sqrt(5.), 0., 0., 1.], \
                [0., -3./np.sqrt(10.), 3./5., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 1./np.sqrt(10.), 4./5., -1./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 3./np.sqrt(10.), 0., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 1./np.sqrt(10.), 1., 1./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., 3./np.sqrt(10.), -3./5., 3./np.sqrt(10.), 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., -1./np.sqrt(10.), 4./5., 1./np.sqrt(10.), 0., 0., 0., 0.], \
                [2./np.sqrt(5.), 3./np.sqrt(10.), 0., 0., 0., 0., 0., -3./np.sqrt(10.),\
                 -2./np.sqrt(5.), 0., 0., 0.], \
                [1./np.sqrt(5.), -1./np.sqrt(10.), 0., 0., -1., 0., 0., -1./np.sqrt(10.),\
                 1./np.sqrt(5.), 0., 0., 0.]])
b51 = np.array([0., 0., 0., 0., 0., -500., 0., -1000., 0., -500., 0., 0.])
x51 = gaussElim(A51.copy(),b51.copy())
print("***ME 581 Homework 2 Print Out***\n")
print("The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, \
F7, F8, F9, FV, FH, FR]T = [")
for it, item in enumerate(x51):
    print("{0:8.7f}, ".format(item), )
print("]")

***ME 581 Homework 2 Print Out***

The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, F7, F8, F9, FV, FH, FR]T = [
6708.2039325, 
3952.8470752, 
-5000.0000000, 
-7115.1247354, 
3500.0000000, 
-7115.1247354, 
-5000.0000000, 
3952.8470752, 
6708.2039325, 
1000.0000000, 
-3000.0000000, 
-2000.0000000, 
]


(b) In this case, $P_3 = 500$, $P_4 = 1000$ and $P_5 = 1500$, the code shown below:

In [31]:
A52 = np.array([[1./np.sqrt(5.), 0., 4./5., 0., 0., 0., 0., 0., 0., 1., 0., 0.],\
                [2./np.sqrt(5.), 0., 3./5., 0., 0., 0., 0., 0., 0., 0., 1., 0.], \
                [0., 0., 0., 0., 0., 0., 3./5., 0., 1./np.sqrt(5.), 0., 0., 0.], \
                [0., 0., 0., 0., 0., 0., 4./5., 0., 2./np.sqrt(5.), 0., 0., 1.], \
                [0., -3./np.sqrt(10.), 3./5., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 1./np.sqrt(10.), 4./5., -1./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 3./np.sqrt(10.), 0., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 1./np.sqrt(10.), 1., 1./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., 3./np.sqrt(10.), -3./5., 3./np.sqrt(10.), 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., -1./np.sqrt(10.), 4./5., 1./np.sqrt(10.), 0., 0., 0., 0.], \
                [2./np.sqrt(5.), 3./np.sqrt(10.), 0., 0., 0., 0., 0., -3./np.sqrt(10.),\
                 -2./np.sqrt(5.), 0., 0., 0.], \
                [1./np.sqrt(5.), -1./np.sqrt(10.), 0., 0., -1., 0., 0., -1./np.sqrt(10.),\
                 1./np.sqrt(5.), 0., 0., 0.]])
b52 = np.array([0., 0., 0., 0., 0., -500., 0., -1000., 0., -1500., 0., 0.])
x52 = gaussElim(A52.copy(),b52.copy())
print("***ME 581 Homework 2 Print Out***\n")
print("The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, \
F7, F8, F9, FV, FH, FR]T = [")
for it, item in enumerate(x52):
    print("{0:8.7f}, ".format(item), )
print("]")

***ME 581 Homework 2 Print Out***

The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, F7, F8, F9, FV, FH, FR]T = [
11068.5364886, 
6561.7261448, 
-7750.0000000, 
-11463.2565181, 
6250.0000000, 
-11463.2565181, 
-8750.0000000, 
5929.2706128, 
11739.3568819, 
1250.0000000, 
-5250.0000000, 
-3500.0000000, 
]


(c) In this case, $P_3 = 1500$, $P_4 = 1000$ and $P_5 = 500$, the code is shown below:

In [32]:
A53 = np.array([[1./np.sqrt(5.), 0., 4./5., 0., 0., 0., 0., 0., 0., 1., 0., 0.],\
                [2./np.sqrt(5.), 0., 3./5., 0., 0., 0., 0., 0., 0., 0., 1., 0.], \
                [0., 0., 0., 0., 0., 0., 3./5., 0., 1./np.sqrt(5.), 0., 0., 0.], \
                [0., 0., 0., 0., 0., 0., 4./5., 0., 2./np.sqrt(5.), 0., 0., 1.], \
                [0., -3./np.sqrt(10.), 3./5., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 1./np.sqrt(10.), 4./5., -1./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 3./np.sqrt(10.), 0., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 1./np.sqrt(10.), 1., 1./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., 3./np.sqrt(10.), -3./5., 3./np.sqrt(10.), 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., -1./np.sqrt(10.), 4./5., 1./np.sqrt(10.), 0., 0., 0., 0.], \
                [2./np.sqrt(5.), 3./np.sqrt(10.), 0., 0., 0., 0., 0., -3./np.sqrt(10.),\
                 -2./np.sqrt(5.), 0., 0., 0.], \
                [1./np.sqrt(5.), -1./np.sqrt(10.), 0., 0., -1., 0., 0., -1./np.sqrt(10.),\
                 1./np.sqrt(5.), 0., 0., 0.]])
b53 = np.array([0., 0., 0., 0., 0., -1500., 0., -1000., 0., -500., 0., 0.])
x53 = gaussElim(A53.copy(),b53.copy())
print("***ME 581 Homework 2 Print Out***\n")
print("The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, \
F7, F8, F9, FV, FH, FR]T = [")
for it, item in enumerate(x53):
    print("{0:8.7f}, ".format(item), )
print("]")

***ME 581 Homework 2 Print Out***

The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, F7, F8, F9, FV, FH, FR]T = [
9056.0753089, 
4506.2456657, 
-7250.0000000, 
-9091.5482730, 
4750.0000000, 
-9091.5482730, 
-6250.0000000, 
5138.7011978, 
8385.2549156, 
1750.0000000, 
-3750.0000000, 
-2500.0000000, 
]


(d) In this case,

In [33]:
A54 = np.array([[1./np.sqrt(5.), 0., 4./5., 0., 0., 0., 0., 0., 0., 1., 0., 0.],\
                [2./np.sqrt(5.), 0., 3./5., 0., 0., 0., 0., 0., 0., 0., 1., 0.], \
                [0., 0., 0., 0., 0., 0., 3./5., 0., 1./np.sqrt(5.), 0., 0., 0.], \
                [0., 0., 0., 0., 0., 0., 4./5., 0., 2./np.sqrt(5.), 0., 0., 1.], \
                [0., -3./np.sqrt(10.), 3./5., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 1./np.sqrt(10.), 4./5., -1./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 3./np.sqrt(10.), 0., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 1./np.sqrt(10.), 1., 1./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., 3./np.sqrt(10.), -3./5., 3./np.sqrt(10.), 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., -1./np.sqrt(10.), 4./5., 1./np.sqrt(10.), 0., 0., 0., 0.], \
                [2./np.sqrt(5.), 3./np.sqrt(10.), 0., 0., 0., 0., 0., -3./np.sqrt(10.),\
                 -2./np.sqrt(5.), 0., 0., 0.], \
                [1./np.sqrt(5.), -1./np.sqrt(10.), 0., 0., -1., 0., 0., -1./np.sqrt(10.),\
                 1./np.sqrt(5.), 0., 0., 0.]])
b54 = np.array([0., 0., 0., 0., 1000., 0., 500., 0., 0., 0., 0., 0.])
x54 = gaussElim(A54.copy(),b54.copy())
print("***ME 581 Homework 2 Print Out***\n")
print("The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, \
F7, F8, F9, FV, FH, FR]T = [")
for it, item in enumerate(x54):
    print("{0:8.7f}, ".format(item), )
print("]")

***ME 581 Homework 2 Print Out***

The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, F7, F8, F9, FV, FH, FR]T = [
4863.4478511, 
1409.8487902, 
-2041.6666667, 
-3755.2047214, 
2541.6666667, 
-4282.2509981, 
-2708.3333333, 
2569.3505989, 
3633.6104634, 
-541.6666667, 
-3125.0000000, 
-1083.3333333, 
]


(e) In this case,

In [34]:
A55 = np.array([[1./np.sqrt(5.), 0., 4./5., 0., 0., 0., 0., 0., 0., 1., 0., 0.],\
                [2./np.sqrt(5.), 0., 3./5., 0., 0., 0., 0., 0., 0., 0., 1., 0.], \
                [0., 0., 0., 0., 0., 0., 3./5., 0., 1./np.sqrt(5.), 0., 0., 0.], \
                [0., 0., 0., 0., 0., 0., 4./5., 0., 2./np.sqrt(5.), 0., 0., 1.], \
                [0., -3./np.sqrt(10.), 3./5., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 1./np.sqrt(10.), 4./5., -1./np.sqrt(10.), 0., 0., 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 3./np.sqrt(10.), 0., -3./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 1./np.sqrt(10.), 1., 1./np.sqrt(10.), 0., 0., 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., 3./np.sqrt(10.), -3./5., 3./np.sqrt(10.), 0., 0., 0., 0.], \
                [0., 0., 0., 0., 0., -1./np.sqrt(10.), 4./5., 1./np.sqrt(10.), 0., 0., 0., 0.], \
                [2./np.sqrt(5.), 3./np.sqrt(10.), 0., 0., 0., 0., 0., -3./np.sqrt(10.),\
                 -2./np.sqrt(5.), 0., 0., 0.], \
                [1./np.sqrt(5.), -1./np.sqrt(10.), 0., 0., -1., 0., 0., -1./np.sqrt(10.),\
                 1./np.sqrt(5.), 0., 0., 0.]])
b55 = np.array([0., 0., 0., 0., 0., 0., -500., 0., -1000., 0., 0., 0.])
x55 = gaussElim(A55.copy(),b55.copy())
print("***ME 581 Homework 2 Print Out***\n")
print("The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, \
F7, F8, F9, FV, FH, FR]T = [")
for it, item in enumerate(x55):
    print("{0:8.7f}, ".format(item), )
print("]")

***ME 581 Homework 2 Print Out***

The member and reaction forces within plane truss are: [F1, F2, F3, F4, F5, F6, F7, F8, F9, FV, FH, FR]T = [
-4863.4478511, 
-1936.8950669, 
2041.6666667, 
3228.1584448, 
-2208.3333333, 
3755.2047214, 
2708.3333333, 
-3096.3968756, 
-3633.6104634, 
541.6666667, 
3125.0000000, 
1083.3333333, 
]
